File: backend/src/test/java/com/skewwhiffy/auraltester/notes/IntervalTests.java
package com.skewwhiffy.auraltester.notes;

import lombok.val;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class IntervalTests {
  @Test
  void canInstantiateMajorSecond() {
    val expected = "major second";

    val actual = Interval.major(2);

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @Test
  void when_diminishingMajorThird_then_minorThird() {
    val expected = "minor third";
    val major = Interval.major(3);

    val actual = major.getDiminished();

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @Test
  void canInstantiateMinorSixth() {
    val expected = "minor sixth";

    val actual = Interval.minor(6);

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @Test
  void when_diminishingMinorSeventh_then_diminishedSeventh() {
    val expected = "diminished seventh";
    val minor = Interval.minor(7);

    val actual = minor.getDiminished();

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @Test
  void when_diminishingDiminishedSecond_then_doublyDiminishedSecond() {
    val expected = "doubly diminished second";
    val diminished = Interval.minor(2).getDiminished();

    val actual = diminished.getDiminished();

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @Test
  void when_diminishingDoublyDiminishedThird_then_3xDiminishedThird() {

    val expected = "3x diminished third";
    val doublyDiminished = Interval.minor(3).getDiminished().getDiminished();

    val actual = doublyDiminished.getDiminished();

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @Test
  void when_augmentingMajorSixth_then_augmentedSixth() {
    val expected = "augmented sixth";
    val major = Interval.major(6);

    val actual = major.getAugmented();

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @ParameterizedTest
  @ValueSource(ints = {1, 4, 5, 8})
  void cannotInstantiateMajorIntervalWhenShouldBePerfect(int degree) {
    assertThatThrownBy(() -> Interval.major(degree)).isInstanceOf(IllegalArgumentException.class);

  }

  @Test
  void canInstantiatePerfectUnison() {
    val expected = "perfect unison";

    val actual = Interval.perfect(1);

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @Test
  void when_diminishingPerfectFourth_then_diminishedFourth() {
    val expected = "diminished fourth";
    val perfect = Interval.perfect(4);

    val actual = perfect.getDiminished();

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @Test
  void when_diminishingDiminishedFifth_then_doublyDiminishedFifth() {
    val expected = "doubly diminished fifth";
    val diminished = Interval.perfect(5).getDiminished();

    val actual = diminished.getDiminished();

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @Test
  void when_diminishingDoublyDiminishedOctave_then_3xDiminishedOctave() {
    val expected = "3x diminished octave";
    val doublyDiminished = Interval.perfect(8).getDiminished().getDiminished();

    val actual = doublyDiminished.getDiminished();

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @Test
  void when_augmentingPerfectUnison_then_augmentedUnison() {
    val expected = "augmented unison";
    val perfect = Interval.perfect(1);

    val actual = perfect.getAugmented();

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @ParameterizedTest
  @ValueSource(ints = {2, 3, 6, 7})
  void cannotInstantiatePerfectIntervalWhenShouldBeMajor(int degree) {
    assertThatThrownBy(() -> Interval.perfect(degree)).isInstanceOf(IllegalArgumentException.class);
  }

  @Test
  void when_augmentingAugmentedSecond_then_doublyAugmentedSecond() {
    val expected = "doubly augmented second";
    val augmented = Interval.major(2).getAugmented();

    val actual = augmented.getAugmented();

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }

  @Test
  void when_augmentingAugmentedFourth_then_3xAugmentedFourth() {
    val expected = "3x augmented fourth";
    val doublyAugmented = Interval.perfect(4).getAugmented().getAugmented();

    val actual = doublyAugmented.getAugmented();

    assertThat(actual.getDisplayString()).isEqualTo(expected);
  }
}

File: backend/src/test/scala/com/skewwhiffy/auraltester/notes/IntervalTests.scala
package com.skewwhiffy.auraltester.notes

import com.skewwhiffy.auraltester.notes.Interval.Interval
import org.scalatest.flatspec.AnyFlatSpec

class IntervalTests extends AnyFlatSpec {
  it should "can instantiate major second" in {
    val expected = "major second"

    val actual = Interval.major(2)

    assert(actual.displayString == expected)
  }

  it should "when diminishing major third then minor third" in {
    val expected = "minor third"
    val major = Interval.major(3)

    val actual = major.diminished

    assert(actual.displayString == expected)
  }

  it should "can instantiate minor sixth" in {
    val expected = "minor sixth"

    val actual = Interval.minor(6)

    assert(actual.displayString == expected)
  }

  it should "when diminishing minor seventh then diminished seventh" in {
    val expected = "diminished seventh"
    val minor = Interval.minor(7)

    val actual = minor.diminished

    assert(actual.displayString == expected)
  }

  it should "when diminishing diminished second then doubly diminished second" in {
    val expected = "doubly diminished second"
    val diminished = Interval.minor(2).diminished

    val actual = diminished.diminished

    assert(actual.displayString == expected)
  }

  it should "when diminishing doubly diminished third then 3x diminished third" in {

    val expected = "3x diminished third"
    val doublyDiminished = Interval.minor(3).diminished.diminished

    val actual = doublyDiminished.diminished

    assert(actual.displayString == expected)
  }

  it should "when augmenting major sixth then augmented sixth" in {
    val expected = "augmented sixth"
    val major = Interval.major(6)

    val actual = major.augmented

    assert(actual.displayString == expected)
  }

  List(1, 4, 5, 8).foreach(degree => {
    it should s"cannot instantiate major $degree" in {
      assertThrows[IllegalArgumentException] {
        Interval.major(degree)
      }
    }
  })

  it should "can instantiate perfect unison" in {
    val expected = "perfect unison"

    val actual = Interval.perfect(1)

    assert(actual.displayString == expected)
  }

  it should "when diminishing perfect fourth then diminished fourth" in {
    val expected = "diminished fourth"
    val perfect = Interval.perfect(4)

    val actual = perfect.diminished

    assert(actual.displayString == expected)
  }

  it should "when diminishing diminished fifth then doubly diminished fifth" in {
    val expected = "doubly diminished fifth"
    val diminished = Interval.perfect(5).diminished

    val actual = diminished.diminished

    assert(actual.displayString == expected)
  }

  it should "when diminishing doubly diminished octave then 3x diminished octave" in {
    val expected = "3x diminished octave"
    val doublyDiminished = Interval.perfect(8).diminished.diminished

    val actual = doublyDiminished.diminished

    assert(actual.displayString == expected)
  }

  it should "when augmenting perfect unison then augmented unison" in {
    val expected = "augmented unison"
    val perfect = Interval.perfect(1)

    val actual = perfect.augmented

    assert(actual.displayString == expected)
  }

  List(2, 3, 6, 7).foreach(degree => {
    it should s"cannot instantiate perfect $degree" in {
      assertThrows[IllegalArgumentException] {
        Interval.perfect(degree)
      }
    }
  })

  it should "when augmenting augmented second then doubly augmented second" in {
    val expected = "doubly augmented second"
    val augmented = Interval.major(2).augmented

    val actual = augmented.augmented

    assert(actual.displayString == expected)
  }

  it should "when augmenting augmented fourth then 3x augmented fourth" in {
    val expected = "3x augmented fourth"
    val doublyAugmented = Interval.perfect(4).augmented.augmented

    val actual = doublyAugmented.augmented

    assert(actual.displayString == expected)
  }

  it should "when equivalent then equal" in {
    def interval = new Interval(6, 69)

    val first = interval
    val second = interval

    assert(first == second)
  }
}